edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+	deviceList      map[string]*v1alpha2.Device
+)
+
+//Start worker
+func (dw DMIWorker) Start() {
+	klog.Infoln("dmi worker start")
+	initDMIActionCallBack()
+	initDeviceModelInfoFromDB()
+	initDeviceInfoFromDB()
+	initDeviceMapperInfoFromDB()
+
+	go dmiserver.StartDMIServer(mapperList, deviceList, deviceModelList)
+
+	for {
+		select {
+		case msg, ok := <-dw.ReceiverChan:
+			if !ok {
+				return
+			}
+
+			if dtMsg, isDTMessage := msg.(*dttype.DTMessage); isDTMessage {
+				if fn, exist := dmiActionCallBack[dtMsg.Action]; exist {
+					err := fn(dw.DTContexts, dtMsg.Identity, dtMsg.Msg)
+					if err != nil {
+						klog.Errorf("DMIModule deal %s event failed: %v", dtMsg.Action, err)
+					}
+				} else {
+					klog.Errorf("DMIModule deal %s event failed, not found callback", dtMsg.Action)
+				}
+			}
+
+		case v, ok := <-dw.HeartBeatChan:
+			if !ok {
+				return
+			}
+			if err := dw.DTContexts.HeartBeat(dw.Group, v); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func initDMIActionCallBack() {
+	dmiActionCallBack = make(map[string]CallBack)
+	dmiActionCallBack[dtcommon.MetaDeviceOperation] = dealMetaDeviceOperation
+}
+
+func dealMetaDeviceOperation(context *dtcontext.DTContext, resource string, msg interface{}) error {
+	message, ok := msg.(*model.Message)
+	if !ok {
+		return errors.New("msg not Message type")
+	}
+	resources := strings.Split(message.Router.Resource, "/")
+	if len(resources) != 3 {
+		return fmt.Errorf("wrong resources %s", message.Router.Resource)
+	}
+	var device v1alpha2.Device
+	var dm v1alpha2.DeviceModel
+	switch resources[1] {
+	case constants.ResourceTypeDevice:
+		switch message.GetOperation() {
+		case model.InsertOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcCreateDevice(&device)
+			if err != nil {
+				klog.Errorf("add device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			deviceList[device.Name] = &device
+		case model.DeleteOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcRemoveDevice(&device)
+			if err != nil {
+				klog.Errorf("delete device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			delete(deviceList, device.Name)
+		case model.UpdateOperation:
+			err := json.Unmarshal(message.Content.([]byte), &device)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+
+			err = GrpcUpdateDevice(&device)
+			if err != nil {
+				klog.Errorf("udpate device %s failed with err: %s", device.Name, err)
+				return err
+			}
+			deviceList[device.Name] = &device
+		default:
+			klog.Warningf("unsupported operation %s", message.GetOperation())
+		}
+	case constants.ResourceTypeDeviceModel:
+		switch message.GetOperation() {
+		case model.InsertOperation:
```

Comment:
these lines is used in all operation, we can move these line above case

Reply:
done